EPE-148 - Fix EOF problem in datastream

jamesfolk-blockone opened the pull request:
This is in reference to this: https://blockone.atlassian.net/browse/EPE-148

* I have added defensive programming practices to cfile to ensure that a user cannot use it if it is closed.

Commit:
added defensive programming practices to cfile to ensure that a user cannot use it if it is closed

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -27,14 +27,16 @@ class cfile_datastream;
 
-   void set_file_path( fc::path file_path ) {
+   void set_file_path( const fc::path &file_path ) {
       _file_path = std::move( file_path );
```
The `std::move` here should be removed if you are going to pass by `const&`. Why was this changed?

jamesfolk-blockone replied:
yes you are right. i'll fix it.

jamesfolk-blockone replied:
it was an attempt to remove the function from calling the copy constructor unnecessarily. I reverted it. since the implementation a move, maybe an r-value param might be better?

Commit:
removed the const reference call

rusty-block-one commented on include/fc/io/cfile.hpp:
```diff
@@ -27,14 +27,16 @@ class cfile_datastream;
    {}
 
-   void set_file_path( fc::path file_path ) {
+   void set_file_path( const fc::path &file_path ) {
```
The std::move of a const& results in a copy of an lvalue. This is no different from stating "_file_path = file_path". It may also confuse the reader into thinking rvalue move semantics are being implemented.

rusty-block-one commented on include/fc/io/cfile.hpp:
```diff
@@ -49,6 +51,9 @@ class cfile {
    ///         "ab+" - open for binary update - create if does not exist
    ///         "rb+" - open for binary update - file must exist
    void open( const char* mode ) {
+       if(_open)
```
For correctness, I would suggest setting _open back to false.

jamesfolk-blockone replied:
good call.

Commit:
for the open function, after it was closing it before an attempt to open it again, I set the internal boolean flag to not open, just in case it might be thrown during the next open attempt

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -27,14 +27,21 @@ class cfile_datastream;
+       if(_open)
+           fclose(_file.get());
+   }
+
```
indentation seems to be 4 spaces instead of 3 spaces as used throughout rest of code. Can you please fix.

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -27,14 +27,21 @@ class cfile_datastream;
 
+   ~cfile() {
+       if(_open)
+           fclose(_file.get());
```
Doesn't this call `fclose` twice on `_file`?
Seems like you could just get rid of the `unique_file` and do it manually since you are anyway.

jamesfolk-blockone replied:
yes. i'm second guessing myself-

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -65,77 +80,118 @@ class cfile {
+        int c = fgetc(_file.get());
+        if(EOF == c)return true;
+
+        return (EOF == ungetc(c, _file.get()));
```
Is this complicated version fo `eof()` needed? Can you provide some information about why these changes are needed. I'm a bit concerned about the `ungetc`:
"A call to fseek, fsetpos or rewind on stream will discard any characters previously put back into it with this function."

jamesfolk-blockone replied:
i was being paranoid. I removed it.

brianjohnson5972 commented on include/fc/io/cfile.hpp:
```diff
@@ -27,14 +27,21 @@ class cfile_datastream;
    }
 
-   fc::path get_file_path() const {
+   const fc::path &get_file_path() const {
```
the pattern we are suppose to follow is "const fc::path& " (don't know if Larry has actually released the coding style document yet, but that is the preferred way)

Commit:
forgot that uptr cleans up the code. removed the extra eof checks

brianjohnson5972 commented on include/fc/io/cfile.hpp:
```diff
@@ -65,77 +80,118 @@ class cfile {
+       if(!_open)
+           throw std::ios_base::failure("cfile is not open");
+
+       if( 0 != fseek( _file.get(), loc, SEEK_SET ) ) {
```
The initial problem talked about EOF issues and had a fix to this method.  Was that determined to not be essential, or was it actually because the file wasn't open?  I would think the open check just prevents getting a confusing error from fseek.

jamesfolk-blockone replied:
I just removed it.

brianjohnson5972 commented on include/fc/io/cfile.hpp:
```diff
@@ -65,77 +80,118 @@ class cfile {
-      int ret = fgetc(_file.get());  
-      if (ret == EOF) {
-         throw std::ios_base::failure( "cfile: " + _file_path.generic_string() +
+        if(feof(_file.get()))return true;
```
move the return to the next line

brianjohnson5972 commented on include/fc/io/cfile.hpp:
```diff
@@ -65,77 +80,118 @@ class cfile {
+        if(feof(_file.get()))return true;
+
+        int c = fgetc(_file.get());
+        if(EOF == c)return true;
```
move the return to the next line

Commit:
addressed code formatting issues

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -49,6 +51,11 @@ class cfile {
+         _open = false;
+      }
+
       _file.reset( FC_FOPEN( _file_path.generic_string().c_str(), mode ) );
```
`_file` is closed twice here. Once above and again with `reset`.  Seems like we either should just go with `_open` flag and raw `fclose` OR remove `_open` flag and just go with `!= nullptr` on `unique_file`.

jamesfolk-blockone replied:
thanks for pointing this out.
it seems that using specifying the delete process for the unique pointer on a FILE, might not be the best of ideas.
Maybe just removing the delete process of the unique pointer might be the better idea. I'll make a PR to show you what i mean-

Commit:
removed the fclose from the unique_ptr cleanup

jamesfolk-blockone commented on include/fc/io/cfile.hpp:
```diff
@@ -25,16 +25,27 @@ class cfile_datastream;
  * std::ios_base::failure exception thrown for errors.
  */
 class cfile {
+private:
```
this code block will ensure that there is only one way to ctor a cfile

swatanabe-b1 commented on include/fc/io/cfile.hpp:
```diff
@@ -15,7 +15,7 @@
 
 namespace detail {
-   using unique_file = std::unique_ptr<FILE, decltype( &fclose )>;
+   using unique_file = std::unique_ptr<FILE>;
```
This is definitely wrong.  You can't `delete` a `FILE*`, which is what the vanilla `unique_ptr` will try to do.

jamesfolk-blockone replied:
right- it seems to me that the unique_ptr is only complicating things.

Commit:
removed the unique_ptr implementation for the FILE*

swatanabe-b1 commented:
I'm not sure what the point of this PR is.  It doesn't address the issue in EPE-148 at all.  Instead it's just changing the usage of unique_ptr which originally worked fine.

jamesfolk-blockone commented:
> I'm not sure what the point of this PR is. It doesn't address the issue in EPE-148 at all. Instead it's just changing the usage of unique_ptr which originally worked fine.

The problem I am trying to fix is a hard to track down EOF issue with cfile. 
With the previous implementation of cfile, when you called the open for the for  the first  time, fclose is called on an undefined  instance of FILE* (unique_ptr::reset())

swatanabe-b1 commented:
> The problem I am trying to fix is a hard to track down EOF issue with cfile.
> With the previous implementation of cfile, when you called the open for the for the first time, fclose is called on an undefined instance of FILE* (unique_ptr::reset())

It isn't undefined, it's null and `fclose` is not called: "if and only if the old value of the stored pointer, old_p, was not equal to nullptr, calls get_deleter()(old_p)" [unique.ptr.single.modifiers].

swatanabe-b1 commented:
The issue with EOF is that `remaining()` may return `true` even if the file position indicator is at the end of the file, because it uses `feof`, but that isn't what `feof` means.

jamesfolk-blockone commented:
> > The problem I am trying to fix is a hard to track down EOF issue with cfile.
> > With the previous implementation of cfile, when you called the open for the for the first time, fclose is called on an undefined instance of FILE* (unique_ptr::reset())
> 
> It isn't undefined, it's null and `fclose` is not called: "if and only if the old value of the stored pointer, old_p, was not equal to nullptr, calls get_deleter()(old_p)" [unique.ptr.single.modifiers].

I see. i did  a quick reference check and that is true-

jamesfolk-blockone commented:
> The issue with EOF is that `remaining()` may return `true` even if the file position indicator is at the end of the file, because it uses `feof`, but that isn't what `feof` means.

Thank you for that. Is that written somewhere? if so, I must have glazed over it.

swatanabe-b1 commented:
It was a comment in the first PR linked in the issue.

jamesfolk-blockone commented:
> It was a comment in the first PR linked in the issue.

I see. Is this the comment from June 8th 2020 from this link? https://github.com/EOSIO/eos/pull/9192

swatanabe-b1 commented:
That is a different, but related issue.  in_avail is another function that is close, but not quite what we need.  The one I was referring to is https://github.com/EOSIO/eos/pull/9183#discussion_r435921222.

jamesfolk-blockone commented:
> That is a different, but related issue. in_avail is another function that is close, but not quite what we need. The one I was referring to is [EOSIO/eos#9183 (comment)](https://github.com/EOSIO/eos/pull/9183#discussion_r435921222).

remaining is a very strange function. is it needed?

swatanabe-b1 commented:
> remaining is a very strange function. is it needed?

Delete it and see what doesn't compile?

swatanabe-b1 commented:
Also, I don't really see it as strange.  The intended behavior is pretty simple: are there any characters left that can be read?

jamesfolk-blockone commented:
> Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?

The reasoning as to why I find it strange is because you could essentially use eof for that use case, instead of a specialized eof wrapped function which has to be maintained.

jamesfolk-blockone commented:
> Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?

it built when i commented out the remaining function

brianjohnson5972 commented:
> > Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?
> 
> it built when i commented out the remaining function

You will need to make sure that all the repos that use it still build, not just EOSIO/eos

jamesfolk-blockone commented:
> > > Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?
> > 
> > 
> > it built when i commented out the remaining function
> 
> You will need to make sure that all the repos that use it still build, not just EOSIO/eos

Ok, Thank you Brian. That is good to know. it sounds easier if i fix the `remaining` function.

Commit:
updated the remaining functipn

Commit:
Fixed  the ``remaining`` function so that it now just checks if the
current getc is EOF.
Added deleted constructors for cfile, cfile_datastream and, datastream.

jamesfolk-blockone commented:
> > > > Also, I don't really see it as strange. The intended behavior is pretty simple: are there any characters left that can be read?
> > > 
> > > 
> > > it built when i commented out the remaining function
> > 
> > 
> > You will need to make sure that all the repos that use it still build, not just EOSIO/eos
> 
> Ok, Thank you Brian. That is good to know. it sounds easier if i fix the `remaining` function.

I believe I have fixed the `remaining` function.

swatanabe-b1 commented on include/fc/io/cfile.hpp:
```diff
@@ -188,13 +227,21 @@ class datastream<fc::cfile, void> : public fc::cfile {
 
    bool get(char& c) {
       c = this->getc();
+      _current_c = c;
```
`get` is not the only function that can affect `remaining`.

swatanabe-b1 commented on include/fc/io/cfile.hpp:
```diff
@@ -188,13 +227,21 @@ class datastream<fc::cfile, void> : public fc::cfile {
+   const datastream& operator=(const datastream &) = delete;
+   const datastream&& operator=(const datastream &&) = delete;
+
+   char _current_c = EOF;
```
EOF is an `int`, not a `char`

jamesfolk-blockone replied:
EOF is defined as `#define	EOF	(-1)`

swatanabe-b1 commented on include/fc/io/cfile.hpp:
```diff
@@ -188,13 +227,21 @@ class datastream<fc::cfile, void> : public fc::cfile {
    }
 
-   bool remaining() { return !this->eof(); }
+   bool remaining() { return (EOF != _current_c); }
```
This still has essentially the same problem as the original.  It checks whether you attempted to read past the end, not whether the datastream is at the end.

jamesfolk-blockone replied:
Thank you  for  pointing that out. I addressed this issue,  now.

Commit:
fixed  the remaining function so that it checks  if there is a byte to read  and is not at  the eof

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -53,10 +53,26 @@ class cfile {
       }
+
+      fseek(_file.get(), 0L, SEEK_END);
+      _size = ftell(_file.get());
```
What if you write to the file after you open it?

jamesfolk-blockone replied:
good catch. thanks.

Commit:
updating the size if there  was a write

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -94,14 +119,24 @@ class cfile {
                                        " unable to write " + std::to_string( n ) + " bytes; only wrote " + std::to_string( result ) );
       }
+      // Keep up with the size  of the file.
+      _size +=  (WRITE_SIZE * n);
```
Do we have a restriction that only one cfile can read/write to a file? I don't think there is that restriction before this change.

jamesfolk-blockone replied:
so what you are saying  is that it  might have been possible for  two instances of cfile to open the same file at the  same time, before these  changes?

jamesfolk-blockone replied:
so what you are saying  is that it  might have been possible for  two instances of cfile to open the same file at the  same time, before these  changes?

heifner replied:
Yes. I don't think there was anything to prevent that use case.

jamesfolk-blockone replied:
ok. that is a good  point. Is this a requirement? Who do I talk to about  this?

jamesfolk-blockone replied:
maybe remaining is  only a requirement for `cfile_datastream`?

heifner replied:
Can't you use the impl you had? 
https://www.geeksforgeeks.org/eof-and-feof-in-c/
https://www.tutorialspoint.com/c_standard_library/c_function_feof.htm

```
    bool eof() const  {
        if(!_open)
            throw std::ios_base::failure("cfile is not open");

        if(feof(_file.get()))return true;

        int c = fgetc(_file.get());
        if(EOF == c)return true;

        return (EOF == ungetc(c, _file.get()));
    }
```

jamesfolk-blockone replied:
good point.

Commit:
removed the current_char as it is not needed anymore

Commit:
removed the size check. fixed remaining function in cfile to work with multiple users

Commit:
removed unused variable

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -53,10 +53,25 @@ class cfile {
+
+      int c(getc());
+      bool isEOF(eof());
+      ungetc(c, _file.get());
```
Check return code for error.

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -53,10 +53,25 @@ class cfile {
+         return false;
+
+      int c(getc());
+      bool isEOF(eof());
```
Really don't like this syntax. Can you at least use `{}` instead of `()` for both of these.

jamesfolk-blockone replied:
no problem. I like to use the initialization list even though it is non-traditional. I will just use the traditional syntax..

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -94,14 +118,20 @@ class cfile {
+      if(!is_open())
+         throw std::ios_base::failure("cfile is not open");
+
+      size_t result = fwrite( d, WRITE_SIZE, n, _file.get() );
```
Where and what is WRITE_SIZE? This needs to be one because `n` is in terms of char.

jamesfolk-blockone replied:
oops- thanks for catching that. i  cleaned up the code with out trying to compile

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -141,6 +182,11 @@ class cfile {
 
 private:
+   cfile(const cfile &) = delete;
+   cfile(const cfile &&) = delete;
```
Should be `cfile( cfile&& ) = delete;`

heifner replied:
Same for other types below.

jamesfolk-blockone replied:
lesson learned*

heifner commented on include/fc/io/cfile.hpp:
```diff
@@ -141,6 +182,11 @@ class cfile {
+   cfile(const cfile &) = delete;
+   cfile(const cfile &&) = delete;
+   const cfile& operator=(const cfile &) = delete;
+   const cfile&& operator=(const cfile &&) = delete;
```
Should be `cfile& operator=( cfile&& ) = delete;`

heifner replied:
Same for others below.

Commit:
fixed a few errors

Commit:
error checking ungetc
